bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,626 @@
+// Note that the Go client libraries don't explicitly have the concept of
+// channel credentials, call credentials, or composite call credentials per
+// [gRPC documentation](https://grpc.io/docs/guides/auth/).
+func getCredentialsOptions(req *pb.CreateClientRequest) ([]grpc.DialOption, error) {
+	var opts []grpc.DialOption
+
+	if req.CallCredential == nil &&
+		req.ChannelCredential == nil &&
+		req.OverrideSslTargetName == "" {
+		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		return opts, nil
+	}
+
+	// If you have call credentials, then you must have channel credentials too
+	if req.CallCredential != nil && req.ChannelCredential == nil {
+		return nil, fmt.Errorf("%s: must supply channel credentials with call credentials", logLabel)
+	}
+
+	// This may not be needed--OverrideSslTargetName is provided to when
+	// creating the channel credentials.
+	if req.OverrideSslTargetName != "" {
+		d := grpc.WithAuthority(req.OverrideSslTargetName)
+		opts = append(opts, d)
+	}
+
+	// Case 1: No additional credentials provided
+	chc := req.GetChannelCredential()
+	if chc == nil {
+		return opts, nil
+	}
+	channelCreds, err := getChannelCredentials(chc, req.OverrideSslTargetName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Case 2: Only channel credentials provided
+	cc := req.CallCredential
+	if cc == nil {
+		d := grpc.WithTransportCredentials(channelCreds)
+		opts = append(opts, d)
+		return opts, nil
+	}
+
+	// Case 3: Both channel & call credentials provided
+	sa := cc.GetJsonServiceAccount()
+	clc, err := oauth.NewJWTAccessFromKey([]byte(sa))
+	if err != nil {
+		return nil, err
+	}
+
+	b := credentialsBundle{
+		channel: channelCreds,
+		call:    clc,
+	}
+
+	d := grpc.WithCredentialsBundle(b)
+	opts = append(opts, d)
+
+	return opts, nil
+}
+
+// getChannelCredentials extracts the channel credentials (credentials for use)
+// with all calls on this client.
+func getChannelCredentials(credsProto *pb.ChannelCredential, sslTargetName string) (credentials.TransportCredentials, error) {
+	var creds credentials.TransportCredentials
+	v := credsProto.GetValue()
+	switch t := v.(type) {
+	case *pb.ChannelCredential_Ssl:
+		pem := t.Ssl.GetPemRootCerts()
+
+		cert, err := x509.ParseCertificate([]byte(pem))
+		if err != nil {
+			return nil, err
+		}
+
+		pool := x509.NewCertPool()
+		pool.AddCert(cert)
+
+		creds = credentials.NewClientTLSFromCert(pool, sslTargetName)
+		if err != nil {
+			return nil, err
+		}
+	case *pb.ChannelCredential_None:
+		creds = insecure.NewCredentials()
+	default:
+		ctx := context.Background()
+		c, err := gauth.FindDefaultCredentials(ctx, "https://www.googleapis.com/auth/cloud-platform")
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO(developer): Determine how to pass this call option back to caller
+		option.WithTokenSource(c.TokenSource)
+
+		return nil, nil
+	}
+	return creds, nil
```

Comment:
You might consider returning out of the case statements above just because this switch statement is a little long. Up to you.

Reply:
The compiler doesn't like it if I remove the final return statement.